test/test_art.py

```diff
@@ -274,6 +275,94 @@ def test_aao_scraper_returns_no_result_when_no_image_present(self):
             next(self.source.get(album, self.settings, []))
 
 
+class ITunesStoreTest(UseThePlugin):
+    def setUp(self):
+        super(ITunesStoreTest, self).setUp()
+        self.source = fetchart.ITunesStore(logger, self.plugin.config)
+        self.settings = Settings()
+        self.album = _common.Bag(albumartist="some artist", album="some album")
+
+    @responses.activate
+    def run(self, *args, **kwargs):
+        super(ITunesStoreTest, self).run(*args, **kwargs)
+
+    def mock_response(self, url, json):
+        responses.add(responses.GET, url, body=json,
+                      content_type='application/json')
+
+    def test_itunesstore_finds_image(self):
+        json = """{
+                    "results":
+                        [
+                            {
+                                "artistName": "some artist",
+                                "collectionName": "some album",
+                                "artworkUrl100": "url_to_the_image"
+                            }
+                        ]
+                  }"""
+        self.mock_response(fetchart.ITunesStore.API_URL, json)
+        candidate = next(self.source.get(self.album, self.settings, []))
+        self.assertEqual(candidate.url, 'url_to_the_image')
+
+    def test_itunesstore_no_result(self):
+        json = '{"results": []}'
+        self.mock_response(fetchart.ITunesStore.API_URL, json)
+        with self.assertRaises(StopIteration):
+            next(self.source.get(self.album, self.settings, []))
+
+    def test_itunesstore_requestexception(self):
+        responses.add(responses.GET, fetchart.ITunesStore.API_URL,
+                      json={'error': 'not found'}, status=404)
+        expected = u'iTunes search failed: 404 Client Error'
+
+        with capture_log('beets.test_art') as logs:
+            with self.assertRaises(StopIteration):
+                next(self.source.get(self.album, self.settings, []))
+
+        self.assertIn(expected, logs[1])
```

Comment:
I like that here you made the choice to not only check that no candidate is yielded, but also that the expected error path is followed.
How do you feel about doing this for the other failure paths?
I have conflicting thoughts about this, on the one hand checking logs for expected messages is not the most elegant or robust -- so much so that people would probably call that integration testing rather than unit testing. On the other hand the way the code is written now, every exception is swallowed up as soon as it is raised and "replaced" by error/debug logging, so there is no easy way to check that the correct failure path is followed without inspecting the logs.

In an ideal world the code would be more modular and there would be individual functions which only return or raise that could be tested individually.

Reply:
Maybe it's a bit over the top, but I've also done it for the other failure paths now.